content/articles/2017-08-tidyr-0.7.0.Rmd

```diff
@@ -0,0 +1,214 @@
+well and will be exported in the next version of dplyr.
+
+```{r}
+# Let's import some tidy eval tools that we'll use in examples below
+library("dplyr")
+sym <- rlang::sym
+```
+
+**Quasiquotation** is essential to program with quoting functions. It
+refers to the ability of unquoting part of a quoted expression, and
+makes it possible to program with the quoting grammars of dplyr and
+tidyr. With quasiquotation, you can change what a function "sees".
+You'll typically want to unquote a symbol representing a data frame
+column with the `!!` operator.
+
+Here, `expand()` sees `vs` and `cyl`:
+
+```{r}
+expand(mtcars, vs, cyl)
+```
+
+Thanks to quasiquotation, we can change what `expand()` sees by
+unquoting the `am` symbol:
+
+```{r}
+x <- sym("am")
+expand(mtcars, vs, !! x)
+```
+
+The second tidy eval feature are **quosures**. A quosure is a special
+kind of expression that evaluates in both the data context (so you can
+refer to data frame columns) and the original context of the
+expression (e.g. a function context, so you can refer to local
+variables created there).
+
+```{r}
+# For nicer printing
+iris <- tibble::as_tibble(iris)
+
+# Let's create a quosure within a local context
+quo <- local({
+  prefix <- "Sepal"
+  quo(starts_with(prefix))
+})
+```
+
+`prefix` only exist in the local context but the quosure can safely
+refer to it
+
+```{r}
+quo
+```
+
+In tidyr 0.7.0, all functions now support quosures:
+
+```{r}
+gather(iris, key, value, !! quo)
+```
+
+Typically you'll use quosures to create wrappers around tidyr
+functions. To this end you'll need `enquo()` which does two things: it
+transforms your function to a dplyr-like quoting function; and it
+returns the quoted expression as a quosure. Creating a wrapper
+function is often a simple matter of enquosing and unquoting:
+
+```{r}
+my_gather <- function(df, expr) {
+  quo <- enquo(expr)
+  tidyr::gather(df, key, value, !! quo)
+}
+```
+
+Thanks to the enquosing, you can safely call your wrapper in local
+contexts (e.g. within a function) and refer to variables defined
+there:
+
+```{r}
+local({
+  prefix <- "Sepal"
+  my_gather(iris, starts_with(prefix))
+})
+```
+
+
+### Special tidy evaluation rules
+
+The tidy evaluation implementation of tidyr is a bit special. While
+the philosophy of tidy evaluation is that R code should refer to real
+objects (from the data frame or from the context), we had to make some
+exceptions to this rule for tidyr. The reason is that several
+functions accept bare symbols to specify the names of _new_ columns to
+create (`gather()` being a prime example). This is not tidy because
+the symbol do not represent any actual object.
+
+Our workaround is to capture these arguments using `rlang::quo_name()`
+(so they still support quasiquotation and you can unquote symbols or
+strings). This type of NSE is now discouraged in the tidyverse:
+symbols in R code should represent real objects.
```

Comment:
Maybe mention that this is not consistent within tidyr - i.e. `separate()` etc use strings. We now prefer strings to refer to variables that don't yet exist.